feat: add shadow variable search post (#102)
mlr-org/mlr3website: mlr-org/gallery/optimization/2023-02-01-shadow-variable-search/index.qmd

As an example, we will search for the optimal set of features for a `r ref("mlr_learners_classif.svm", "support vector machine")` on the `r ref("mlr_tasks_pima", text = "Pima Indian Diabetes")` data set.
We focus on the `r ref("mlr_fselectors_shadow_variable_search", "shadow variable search")` algorithm and the application with the `r ref("mlr3fselect::fselect()")` function.
We assume that you are already familiar with the basic building blocks of the mlr3 ecosystem.
If you are new to feature selection, we recommend reading the [feature selection chapter](https://mlr3book.mlr-org.com/feature-selection.html) of the mlr3book.
Some knowledge about `r ref_pkg("mlr3pipelines")` is beneficial but not necessary to understand the example.

# Shadow Variable Search

Adding shadow variables to a data set is a commonly used method in machine learning [@wu_controlling_2007; @thomas_probing_2017].
The idea is to add permutated copies of the original features to the data set.
These permutated copies are called shadow variables or pseudovariables and the permutation makes them unrelated to the target variable.
The subsequent search is similar to the sequential forward selection algorithm.

> Features are added to the model one at a time based on the highest performance improvement.

Reviewer: ```suggestion
This new feature is selected as the one that improves the performance of the model the most. 
This selection is computationally expensive, as one model for each of the not yet included features has to be trained.
```

> The difference is that the termination criterion is the selection of a shadow variable.

Reviewer: ```suggestion
The difference between *Shadow Variable Search* and *Sequential Forward Selection* is that the former uses the selection of a shadow variable as the termination criterion.
```

Selecting a shadow variable means that the best improvement is achieved by adding a feature that is unrelated to the target variable.

> Consequently, the variables not yet selected are also correlated to the target variable only by chance.

Reviewer: ```suggestion
```

Reviewer: Die Aussage ist zu stark finde finde ich (man erwartet das zwar schon irgendwie aber ich würde dafür nicht meine Beine ins Feuer legen)

Author: Ich füg mal most likely hinzu

> Therefore, only the previously selected features have a true influence on the target variable.

Reviewer: ```suggestion
```

Author: Also ganz rausnehmen würde ich das nicht. Das ist halt die Annahme von dem Algorithmus. Ob das bei allen Datensätzen und Learner zutrifft, ist natürlich fraglich.

Feature selection algorithms are implemented with the `r ref("mlr3fselect::FSelector")` class.

> A complete list of all fselectors is available on the [website](https://mlr-org.com/fselectors.html).

Reviewer: ```suggestion
A complete list of all fselectors is available on the [mlr website](https://mlr-org.com/fselectors.html).
```

> We load the `r ref("mlr_fselectors_shadow_variable_search", "shadow variable search")` with the `r ref("mlr3fselect::fs()")` function.

Reviewer: ```suggestion
```

Reviewer: das finde ich da irgendwie random

Reviewer: also weiter unten habe ich es jetzt hingeschrieben

> The algorithm has no parameters.

Reviewer: ```suggestion
```

The data set includes 768 patients with 8 measurements.

```{r 2023-02-01-shadow-variable-search-003}
task = tsk("pima")
```

```{r 2023-02-01-shadow-variable-search-004}
#| code-fold: true
#| column: page
#| fig-width: 14
#| fig-height: 3
#| fig.cap: "Distribution of the features in the Pima Indian Diabetes data set."
#| warning: false

library(ggplot2)
library(data.table)

task = tsk("pima")
data = melt(as.data.table(task), id.vars = task$target_names, measure.vars = task$feature_names)

ggplot(data, aes(x = value, fill = diabetes)) +
  geom_density(alpha = 0.5) +
  facet_wrap(~ variable, ncol = 8, scales = "free") +
  scale_fill_viridis_d(end = 0.8) +
  theme_minimal() +
  theme(axis.title.x = element_blank())
```

The data set contains missing values.

```{r 2023-02-01-shadow-variable-search-005}
task$missings()
```

Support vector machines cannot handle missing values.
We impute the missing values with the `r ref("mlr_pipeops_imputehist", "histogram imputation")` method.

```{r 2023-02-01-shadow-variable-search-006}
learner = po("imputehist") %>>% lrn("classif.svm", predict_type = "prob")
```

# Feature Selection

> We can now start the feature selection directly with the `r ref("fselect()")` function.

Reviewer: ```suggestion
We can now start the feature selection directly with the `r ref("fselect()")` function.
To set the method to `r ref("mlr_fselectors_shadow_variable_search", "shadow variable search")`, we use the  `r ref("mlr3fselect::fs()")` function.
A complete list of all fselectors is available on the [mlr website](https://mlr-org.com/fselectors.html).
```

> In addition to the task, learner and feature selection algorithm, we have to select a `r ref("Resampling", text = "resampling strategy")` and `r ref("Measure", text = "performance measure")` to determine how the performance of a feature subset is evaluated.

Reviewer: ```suggestion
In addition to the task, learner and method, we have to select a `r ref("Resampling", text = "resampling strategy")` and `r ref("Measure", text = "performance measure")` to determine how the performance of a feature subset is evaluated.
```

```{r 2023-02-01-shadow-variable-search-007}
instance = fselect(
  method = fs("shadow_variable_search"),
  task = task,
  learner = learner,
  resampling = rsmp("cv", folds = 3),
  measures = msr("classif.auc")
)
```

> The `r ref("fselect()")` function returns an `r ref("mlr3fselect::FSelectInstanceSingelCrit")` that includes an archive with all evaluated feature sets.

Reviewer: Also ich finde es irgendwie verwirrend das archive hier zu zeigen. Entweder nicht zeigen oder ein bissl erklären. 
Also in der Algorithmus-Beschreibung oben, ist ja eine iteration: 
"Fitte ein Model für jedes neue Feature und füge das hinzu, das am meisten bringt". 
Die Darstellung im Archive ist ja viel detaillierter weil halt jedes Model dargestellt wird.

Author: Okay also zeige ich lieber den Optimization Path
